flashbax/vault/vault.py

```diff
@@ -0,0 +1,448 @@
+
+
+class Vault:
+    def __init__(
+        self,
+        vault_name: str,
+        experience_structure: Optional[Experience] = None,
+        rel_dir: str = "vaults",
+        vault_uid: Optional[str] = None,
+        metadata: Optional[dict] = None,
+    ) -> None:
+        """Flashbax utility for storing buffers to disk efficiently.
+
+        Args:
+            vault_name (str): the upper-level name of this vault.
+                Resulting path is <cwd/rel_dir/vault_name/vault_uid>.
+            experience_structure (Optional[Experience], optional):
+                Structure of the experience data, usually given as `buffer_state.experience`.
+                Defaults to None, which can only be done if reading an existing vault.
+            rel_dir (str, optional):
+                Base directory of all vaults. Defaults to "vaults".
+            vault_uid (Optional[str], optional): Unique identifier for this vault.
+                Defaults to None, which will use the current timestamp.
+            metadata (Optional[dict], optional):
+                Any additional metadata to save. Defaults to None.
+
+        Raises:
+            ValueError:
+                If the targeted vault does not exist, and no experience_structure is provided.
+
+        Returns:
+            Vault: a vault object.
+        """
+        # Get the base path for the vault and the metadata path
+        vault_str = vault_uid if vault_uid else datetime.now().strftime("%Y%m%d%H%M%S")
+        self._base_path = os.path.join(os.getcwd(), rel_dir, vault_name, vault_str)
+        metadata_path = epath.Path(os.path.join(self._base_path, METADATA_FILE))
+
+        # Check if the vault exists, otherwise create the necessary dirs and files
+        base_path_exists = os.path.exists(self._base_path)
+        if base_path_exists:
+            print(f"Loading vault found at {self._base_path}")
+            # Vault exists, so we load the metadata to access the structure etc.
+            self._metadata = json.loads(metadata_path.read_text())
+
+            # Ensure minor versions match
+            assert (self._metadata["version"] // 1) == (VERSION // 1)
+
+        elif experience_structure is not None:
+            print(f"New vault created at {self._base_path}")
+
+            # Create the necessary dirs for the vault
+            os.makedirs(self._base_path)
+
+            # Ensure provided metadata is json serialisable
+            metadata_json_ready = jax.tree_util.tree_map(
+                lambda obj: str(obj)
+                if not isinstance(obj, (bool, str, int, float, type(None)))
+                else obj,
+                metadata,
+            )
+
+            # We save the structure of the buffer state
+            #  e.g. [(128, 100, 4), jnp.int32]
+            # We will use this structure to map over the data stores later
+            serialised_experience_structure = jax.tree_map(
+                lambda x: [str(x.shape), x.dtype.name],
```

Comment:
I think I found a solution for this issue of having to check for lists/not allowing lists in the experience, jax has a ShapeDtypeStruct: https://jax.readthedocs.io/en/latest/_autosummary/jax.ShapeDtypeStruct.html

I think you'd just need to make a `serialize_shape_dtype` and `deserialize_shape_dtype`

Reply:
Took an alternative approach of saving the shape and dtype metadata in separate pytrees 🖖🏻